spmd/tensor/device_mesh.py

```diff
@@ -69,61 +95,114 @@ def __init__(self, device_type, mesh):
                 f"DeviceMesh only support cpu or cuda device type, but got {device_type}"
             )
 
-        # TODO: support multi-dimensional device mesh
-        assert self.mesh.ndim == 1, "Only support 1-d device mesh for now"
+        world_size = get_world_size()
+        if self.mesh.numel() > world_size:
+            raise RuntimeError(
+                f"Mesh should not be bigger than default world size, but found {self.mesh.numel()} ranks!"
+            )
 
-    # pyre-fixme[3]: Return type must be annotated.
-    def __enter__(self):
+        unique_mesh_values = self.mesh.unique(sorted=True)
+        if unique_mesh_values.numel() != self.mesh.numel():
+            raise RuntimeError(
+                f"DeviceMesh cannot have duplicate values, but found {self.mesh.tolist()}"
+            )
+
+        if dim_groups is not None:
+            # if user hand creating dimension based groups
+            # we just take it and use it for communication
+            # we assume user passing dim_gruops are legit
+            # TODO: add more checks to check the correctness
+            # of user passed in dim groups
+            self._dim_to_groups = dim_groups
+            return
+
+        self._dim_to_groups: List[List[ProcessGroup]] = [
+            [] for _ in range(self.mesh.ndim)
+        ]
+        # create sub pgs base on the mesh argument
+        for dim in range(self.mesh.ndim):
```

Comment:
So if I understand this correctly, if we have 2D mesh, then we will have 2 sub-process groups right?